### PR TITLE
B-Tax UI at /btax

### PR DESCRIPTION
**Pull request by PeterDSteinberg**
This is a work-in-progress for a user interface for B-Tax.  See attached screenshot below:

[Welcome to TaxBrain _ TaxBrain.pdf](https://github.com/OpenSourcePolicyCenter/webapp-public/files/379134/Welcome.to.TaxBrain._.TaxBrain.pdf)

### CODE REVIEW

**Comment by talumbau** [comment]
cool, thanks @PeterDSteinberg. @jdebacker and @bfgard please have a look at the attached PDF and see if this aligns with your ideas on interface.

**Comment by jdebacker** [comment]
Thanks @PeterDSteinberg!  I think the layout looks good.  Would it be possible to consolidate the depreciation items into a table in order to condense things a bit?  The same options are available for each of the variously lived assets, so there is a lot of repetition here.  It might also make sense to allow the user to apply changes to all classes easily (e.g. switch to expensing of investment for all assets could be done with one click, rather than having to do it for each class individually).  

@MattHJensen any thoughts?

**Comment by PeterDSteinberg** [comment]
@jdebacker I can work on consolidation of the deprecation asset year parameter groups into a table-like structure with a toggle on check/uncheck a given parameter for all years.

**Comment by MattHJensen** [comment]
> Would it be possible to consolidate the depreciation items into a table in order to condense things a bit? The same options are available for each of the variously lived assets, so there is a lot of repetition here. It might also make sense to allow the user to apply changes to all classes easily (e.g. switch to expensing of investment for all assets could be done with one click, rather than having to do it for each class individually).

> @MattHJensen any thoughts?

Makes sense to me too.

**Comment by PeterDSteinberg** [comment]
This commit converts the Deprecation section of B-Tax to more compact table with a row to check-all.  See below:

screen shot 2016-07-25 at 12 23 42 pm

On-click actions like checking all rows in a column have not been added yet.

**Comment by jdebacker** [comment]
@PeterDSteinberg , this looks good!  Seeing this now, let me propose a few small changes:

1. Change "Asset year life" to "Property class"
2. Change "Check all asset years" to "Check all property classes" or just "Check all"
3.  Make the last column the 4th and retitle- so columns 2-4 are GDS, ADS, Economic ("Economic" will replace "Tax=Economic?").
4. Make the Economic radio button a round one like the GDS and ADS.
5. Ensure that only one of the radio buttons in the first 3 columns (GDS, ADS, Economic) can be selected at one time.
6. Let's replace the "full expensing" column to one that reads "Bonus" and has fields to enter values rather than radio buttons.  The mouse over info button will alert the user that this is the bonus deprecation rate and that if one wants to move to full expensing, they will enter a rate of 100% in this field.

Let me know if you have questions.  Thanks.

**Comment by PeterDSteinberg** [comment]
@jdebacker So on item in 6) in your list above, do you mean that the to-be-renamed Bonus column should have just a single text box for a percentage?  Is 100% the default there?

**Comment by jdebacker** [comment]
Yes.  The default will be 50% for 20 year and less property.  0% for longer-lived property.

**Comment by PeterDSteinberg** [comment]
The latest commit reorganizes the table based on the discussion in the last couple comments, @jdebacker .  There is still work to be done on getting the right defaults in the table and on-click actions.  The last commit also removes a lot of the unused code that I had cut and pasted over from `/taxbrain` (except fro views.py which still needs to be cleaned up). Here's how it looks now:
screen shot 2016-07-25 at 9 25 19 pm

**Comment by jdebacker** [comment]
@PeterDSteinberg thanks.  This looks good to me.

**Comment by PeterDSteinberg** [comment]
With the new parameter jsons from @jdebacker, this is what it looks like now:

[Welcome to TaxBrain _ TaxBrain.pdf](https://github.com/OpenSourcePolicyCenter/webapp-public/files/391462/Welcome.to.TaxBrain._.TaxBrain.pdf)

**Comment by PeterDSteinberg** [comment]
Currently with last commit's code I am getting an error when I try to create a separate class from `OutputUrl` that is like an `OutputUrl` for B-Tax.  I cannot run `migrate` now because of this problem.  It seems to relate to `pk` (primary key).

```
$ python manage.py migrate -v 3
Operations to perform:
  Synchronize unmigrated apps: gunicorn, staticfiles, messages, humanize, import_export, storages
  Apply all migrations: account, sessions, admin, flatblocks, register, dynamic, sites, auth, taxbrain, contenttypes, btax, hermes, pages
Synchronizing apps without migrations:
Running pre-migrate handlers for application admin
Running pre-migrate handlers for application auth
Running pre-migrate handlers for application contenttypes
Running pre-migrate handlers for application sessions
Running pre-migrate handlers for application sites
Running pre-migrate handlers for application taxbrain
Running pre-migrate handlers for application dynamic
Running pre-migrate handlers for application pages
Running pre-migrate handlers for application register
Running pre-migrate handlers for application btax
Running pre-migrate handlers for application flatblocks
Running pre-migrate handlers for application account
Running pre-migrate handlers for application hermes
Running pre-migrate handlers for application import_export
  Creating tables...
    Running deferred SQL...
  Installing custom SQL...
Loading 'initial_data' fixtures...
Checking '/Users/psteinberg/Documents/AEI/webapp-public' for fixtures...
No fixture 'initial_data' in '/Users/psteinberg/Documents/AEI/webapp-public'.
Loading 'initial_data' fixtures...
Checking '/Users/psteinberg/Documents/AEI/webapp-public' for fixtures...
No fixture 'initial_data' in '/Users/psteinberg/Documents/AEI/webapp-public'.
Loading 'initial_data' fixtures...
Checking '/Users/psteinberg/Documents/AEI/webapp-public' for fixtures...
No fixture 'initial_data' in '/Users/psteinberg/Documents/AEI/webapp-public'.
Loading 'initial_data' fixtures...
Checking '/Users/psteinberg/Documents/AEI/webapp-public' for fixtures...
No fixture 'initial_data' in '/Users/psteinberg/Documents/AEI/webapp-public'.
Loading 'initial_data' fixtures...
Checking '/Users/psteinberg/Documents/AEI/webapp-public' for fixtures...
No fixture 'initial_data' in '/Users/psteinberg/Documents/AEI/webapp-public'.
Loading 'initial_data' fixtures...
Checking '/Users/psteinberg/Documents/AEI/webapp-public' for fixtures...
No fixture 'initial_data' in '/Users/psteinberg/Documents/AEI/webapp-public'.
Running migrations:
  Rendering model states...Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/core/management/__init__.py", line 338, in execute_from_command_line
    utility.execute()
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/core/management/__init__.py", line 330, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/core/management/base.py", line 390, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/core/management/base.py", line 441, in execute
    output = self.handle(*args, **options)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 221, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/migrations/executor.py", line 104, in migrate
    state = migration.mutate_state(state, preserve=do_run)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/migrations/migration.py", line 83, in mutate_state
    operation.state_forwards(self.app_label, new_state)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/migrations/operations/fields.py", line 51, in state_forwards
    state.reload_model(app_label, self.model_name_lower)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/migrations/state.py", line 152, in reload_model
    self.apps.render_multiple(states_to_be_rendered)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/migrations/state.py", line 262, in render_multiple
    model.render(self)
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/migrations/state.py", line 546, in render
    body,
  File "/Users/psteinberg/miniconda3/envs/webapp/lib/python2.7/site-packages/django/db/models/base.py", line 254, in __new__
    'base class %r' % (field.name, name, base.__name__)
django.core.exceptions.FieldError: Local field u'id' in class 'OutputUrl' clashes with field of similar name from base class 'OutputUrl'
```

File: webapp/apps/btax/urls.py [hunk 1]
```diff
@@ -4,10 +4,10 @@
 
 
 urlpatterns = patterns('',
-    url(r'^$', btax_results, name='tax_form'),
-    url(r'^(?P<pk>\d+)/output.csv/$', csv_output, name='csv_output'),
-    url(r'^(?P<pk>\d+)/input.csv/$', csv_input, name='csv_input'),
-    url(r'^(?P<pk>\d+)/', output_detail, name='output_detail'),
+    url(r'^$', btax_results, name='btax_tax_form'),
+    url(r'^(?P<pk>\d+)/output.csv/$', csv_output, name='btax_csv_output'),
```
Comment (PeterDSteinberg) [inline webapp/apps/btax/urls.py]: Are these `pk` urls distinct from those in `webapp/apps/taxbrain/urls.py`?

File: webapp/apps/btax/models.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+from uuidfield import UUIDField
+from jsonfield import JSONField
+import datetime
+
+from ..taxbrain.models import (SeparatedValuesField as _SeparatedValuesField,
+                               CommaSeparatedField as _CommaSeparatedField,
+                               convert_to_floats)
+
+class SeparatedValuesField(_SeparatedValuesField):
```
Comment (talumbau) [inline webapp/apps/btax/models.py]: I'm confused regarding `SeparatedValuesField` and `CommaSeparatedField`. Why not just import them and inherit directly?

File: webapp/apps/btax/urls.py (hunk 1 above)
Comment (talumbau) [inline webapp/apps/btax/urls.py]: yes, if you set up all of the `urls.py` correctly, someone would do ospc.org/taxbrain/342 and get the OutputURL with that ID, This has a key to the model that holds user input and results. In this case, we probably want to set up \<host\>/btax/\<number\> for the results page, at least for now.

**Comment by talumbau** [comment]
This SO post:
http://stackoverflow.com/questions/2344751/in-django-model-inheritance-does-it-allow-you-to-override-a-parent-models-a

leads me to believe the error might be due to the unusual inheritance model you are using with the class I mentioned above. I'll keep digging but just a heads up.

**Comment by PeterDSteinberg** [comment]
@talumbau  I fixed the problem by removing the inheritance and also `rm -rf` my migrations folders.

Now it has place holders for running btax.  You can change a parameter, then run, then wait, then get redirected to the tables page, but the tables are not existing there yet.  I'll see what it takes to get this PR working with the two B-Tax PRs: 
 * [https://github.com/open-source-economics/B-Tax/pull/40](https://github.com/open-source-economics/B-Tax/pull/40) and 
 * [https://github.com/open-source-economics/B-Tax/pull/38](https://github.com/open-source-economics/B-Tax/pull/38)

**Comment by talumbau** [comment]
Hmm.. we definitely don't want to remove existing migrations, so the removal of those migrations should not be part of the PR. But I'm glad it's working. The other thing I noticed in this PR is that the number of files is quite large and it seems like larger than is necessary. I will comment on certain files but we might need to do a screenshare so I can understand the rationale for some of the files

File: static/css/taxbrain.css [hunk 3]
```diff
@@ -8440,6 +8440,18 @@ label {
   border-right: 2px solid #dddddd;
   padding-left: 30px;
 }
+.table-check-all {
```
Comment (talumbau) [inline static/css/taxbrain.css]: What's the reasoning behind these changes?

File: templates/btax/header.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{% load staticfiles %}
```
Comment (talumbau) [inline templates/btax/header.html]: this header is identical to the standard taxbrain header, is it not? Shouldn't we just import that template instead?

File: templates/btax/includes/params/inputs/field.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{% load utility %}
```
Comment (talumbau) [inline templates/btax/includes/params/inputs/field.html]: this seems like a duplication of the existing field.html. Is that correct?

File: templates/btax/includes/params/inputs/param.html [hunk 6]
```diff
@@ -0,0 +1,26 @@
+{% load inputs %}
```
Comment (talumbau) [inline templates/btax/includes/params/inputs/param.html]: Is this a duplicate of the existing param.html?

File: templates/btax/header.html (hunk 4 above)
Comment (PeterDSteinberg) [inline templates/btax/header.html]: yes.  this all started from a big cut and paste.  I'll reference existing file instead.

File: templates/btax/includes/params/inputs/field.html (hunk 5 above)
Comment (PeterDSteinberg) [inline templates/btax/includes/params/inputs/field.html]: same comment as for header.html

File: templates/btax/includes/params/tooltip.html [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{% if text %}
```
Comment (talumbau) [inline templates/btax/includes/params/tooltip.html]: this looks to be a duplicate of the existing tooltip.html

File: templates/btax/includes/params/used/amt.html [hunk 8]
```diff
@@ -0,0 +1,45 @@
+<div class="parameter-block">
+  <div class="parameter-block-header">
```
Comment (talumbau) [inline templates/btax/includes/params/used/amt.html]: this looks to be a duplicate of the existing amt.html

File: static/css/taxbrain.css (hunk 3 above)
Comment (PeterDSteinberg) [inline static/css/taxbrain.css]: I had intended to use them in the row at the top of the table in depreciation of btax (the row that lets you check all).  I'll make sure it is still referenced.

Comment (talumbau) [inline static/css/taxbrain.css]: I'm not much of a css person, but I wanted to point out that my understanding is that the CSS is generated from the `less` files. Would it make sense to add to the appropriate `less` file and then have the new version of `taxbrain.css` generated from that?

Comment (PeterDSteinberg) [inline static/css/taxbrain.css]: Will fix.  I haven't used `less` before - didn't realize there was a pre-build step.

Comment (talumbau) [inline static/css/taxbrain.css]: yes, this was new to me as well. See the conversation with @mistakevin on Tax-Calculator Flowdock

Comment (mistakevin) [inline static/css/taxbrain.css]: I can move bootstrap to `package.json` if that will help simplify things up. That way we won't have to do `bower install` in addition to `npm install`

File: templates/btax/includes/params/inputs/param.html (hunk 6 above)
Comment (PeterDSteinberg) [inline templates/btax/includes/params/inputs/param.html]: Yes.  Removed.

**Comment by PeterDSteinberg** [comment]
TODO:
 * [ ] logos (404 currently)
 * [x] remove unused code that was cut and pasted over from /webapp/apps/taxbrain to webapp/apps/btax
 * [x] fix the checkbox and radio button styling (put them all in one row) for the different asset year blocks within the Deprecation section
 * [ ] tests similar to those in webapp/apps/taxbrain/tests
 * [x] logic related to POST requests (so far this is just a UI that supports GET requests)
 * [ ] results.html work in Bootstrap.  Here is an example of the table data that will be accessible to Bootstrap when making results tables: 
[btax_table_format_check.txt](https://github.com/OpenSourcePolicyCenter/webapp-public/files/415707/btax_table_format_check.txt)
* [ ] after the work on results.html, see if more template files in templates/btax can be removed in favor of using templates/taxbrain items, such as not_ready.html.

**Comment by talumbau** [comment]
@PeterDSteinberg one thing to mention about TravisCI on this repo: we have a few private environment variables set for testing. The private environment variables are only available in Travis CI for PRs where the originating branch is on this repo. So your PR that is coming from your fork won't have the environment variables. One way to fix this is to push your branch to this repo and do the PR that way, or I can push your branch to this repo, although for ease of development we should probably make sure you can push a branch to this repo. Let me know what happens when you try it.

**Comment by PeterDSteinberg** [comment]
@talumbau Looks like I don't have write access to this repo (just tried to push here).

**Comment by talumbau** [comment]
ok, I'll work on this. Shouldn't take too long.

**Comment by PeterDSteinberg** [comment]
Closing.  Re-opened as #304

**Pull request closed without merging**